File: python/questions/buildTree1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        inm = {}
        for p in range(len(inorder)):
            inm[inorder[p]] = p
        def build(ine: int, poste: int, len: int) -> Optional[TreeNode]:
            if len == 0:
                return None
            ip = inm[postorder[poste]]
            rl = ine - ip
            return TreeNode(postorder[poste], build(ip-1, poste-rl-1, len-rl-1), build(ine, poste-1, rl))
        return build(len(inorder)-1, len(postorder)-1, len(inorder))
            

File: c/questions/buildTree1.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* build(int* inm, int* inorder, int ine, int* postorder, int poste, int len) {
    if (len == 0) {
        return NULL;
    }
    int ip = inm[postorder[poste]+3000];
    int rl = ine - ip;
    struct TreeNode* node = (struct TreeNode*) malloc(sizeof(struct TreeNode));
    node->val = postorder[poste];
    node->left = build(inm, inorder, ip-1, postorder, poste-rl-1, len-rl-1);
    node->right = build(inm, inorder, ine, postorder, poste-1, rl);
    return node;
}
    
struct TreeNode* buildTree(int* inorder, int inorderSize, int* postorder, int postorderSize) {
    int* inm = (int*) malloc(sizeof(int) * 6000);
    for (int p=0; p<inorderSize; p++) {
        inm[inorder[p]+3000] = p;
    }
    return build(inm, inorder, inorderSize-1, postorder, postorderSize-1, inorderSize);
}
